omskriving af introduction to react (#77)
SW705E19/report: sections/design/introduction-to-react.tex

\end{lstlisting}
The variable that is embedded in the JSX can be any valid JavaScript expression \cite{introducingJSX}.

> \\\\

Reviewer: ```suggestion
```

export default Button
\end{lstlisting}
As an example we can create a button component with an on-click event embedded into it.
Here we use brackets to enable us to embed a callback function in the JSX which will be called when the button is clicked.
We can now export this button and it can be imported by other components.
There are both class components and functional components.
An example of this can be seen in the following listings:

\begin{lstlisting}
function Button(props) {
	return (
		<div>
			<button onClick={console.log(props.textToPrint)}>Hello React</button>
		</div>
	);
}
\end{lstlisting}

> We can convert this into a class like this:

Reviewer 2: ref

Author: har pushet hvor jeg referer til dem begge to ovenover. Men skal jeg også referere til den i linjen imellem dem ?

\subsection*{State}
There two kinds of components, which are stateful and stateless components.
In our application we will mostly implement stateful components as class components and stateless components as functional components. 
Functional components can also be stateful, but the choice of using class components is mostly a question of preference.
A stateful component has a state which contains all the dynamic data in the component.
This is useful since the render method is called every time the local \textit{State} is changed.
We can set the initial state of a component by assigning \texttt{this.state} in the constructor of the component.
\begin{center}
    \texttt{this.state = \{textToDisplay: 'Initial text'\};}
\end{center}
The state can then be changed by reassigning to the \texttt{date} variable.
This is done by using the \texttt{this.setState()} function.
\begin{center}
    \texttt{this.setState(\{textToDisplay: 'New text'\});}
\end{center}
Parent or child components should not care if a certain component is stateful or stateless since the state is local and not accessible by other components.

> A component can however pass variables from its state through \texttt{props} to a child component, which means that information about a state can only be passed in a \textit{top-down} approach \cite{ReactJS}.

Reviewer 2: hvorfor TT på props og it på top-down?

Every time a state is changed with \texttt{this.setState()} the component that the state belongs to and all of its children are re-rendered.
By passing callback to the props of children we can allow  the children of a component to transfer data back to its parent and even trigger re-renders by passing callback functions that set the state of the parent component.
If a function calls \texttt{this.setState()}, it has to be bound in the constructor of the component.
An example of how this is done can be seen in the following listing.
\begin{lstlisting}
class ourComponent extends Component {
	constructor(props){
		super(props);
		this.state = {
			textToDisplay: 'initial text'
		};

		this.handleChange = this.handleChange.bind(this);
	}

	handleChange(text){
		this.setState({
			textToDisplay: text
		});
	}

	render(){
		return <ChildComponent handleChange={this.handleChange} />
	}
}
\end{lstlisting}

\subsection*{Lifecycle}

> When a component is implemented as a class that extends \texttt{Component} from the react library, it gives us access to some functions that are automatically called at different points in the lifecycle of a component.

Reviewer 2: ```suggestion
When a component is implemented as a class that extends \texttt{Component} from the React library, it gives us access to some functions that are automatically called at different points in the lifecycle of a component.
```

This function is called right after the component is rendered for the first time in its lifecycle.
This is useful for when we need to fetch data from the API that we want to save in the state of the component.
The reason that this is not done in the constructor of the component is that the render function should not be waiting for asynchronous operations before it is allowed to render.
This is because we do not know how long it will take before the asynchronous function is done executing, or if it will even succeed.
Instead of waiting, it is better to first render the component before calling asynchronous functions that will trigger a re-rendering of the component.
\begin{lstlisting}[caption={Example with the componentDidMount function}, label={lst:componentDidMount-example}]
class OurComponent extends Component {
	constructor(props){
		super(props);
		this.state = {
			data: null
		};
	}

	componentDidMount(){
		userService.getUserById(1).then((dataRes) => {
			this.setState({
				data: dataRes
			});
		})
	}

	render(){
		if(this.state.data !== null){
			return <ChildComponent data={this.state.data} />:
		}
			
		return null;
	}
}
\end{lstlisting}

> An example of \texttt{componentDidMount} can be seen on \autoref{lst:componentDidMount-example}.

Reviewer: Tænker nok det er bedst at have reffen ovenover listing

We initialize the state with a null.
We check in the render function if the value in the state is null.
If it is null we return null, which just renders nothing.
If it is not null we render a component and pass the data as a prop.
The first time the render function is called, the value is null.
After it has rendered the \texttt{componentDidMount} function is called.

> In \texttt{componentDidMount} we make an API call and then set \texttt{data} in the \texttt{state} equals to the response.

Reviewer: ```suggestion
In \texttt{componentDidMount} we make an API call and then set \texttt{data} in the \texttt{state} equal to the response.
```

export default Button
\end{lstlisting}
As an example we can create a button component with an on-click event embedded into it.
Here we use brackets to enable us to embed a callback function in the JSX which will be called when the button is clicked.
We can now export this button and it can be imported by other components.
There are both class components and functional components.

> An example of this can be seen in the following listings:

Reviewer: tilføj labels og referer til de pågældende

\begin{lstlisting}
function Button(props) {
	return (
		<div>
			<button onClick={console.log(props.textToPrint)}>Hello React</button>
		</div>
	);
}
\end{lstlisting}

We can convert this into a class like this:

\begin{lstlisting}
class Button extends React.Component {
	constructor(props){
		super(props);
	}

	render() {
		return (
			<div>
				<button onClick={console.log('Hello')}>Hello React</button>
			</div>
		);
	}
}
\end{lstlisting}

A component is conceptually a JavaScript function or class, which means that they can accept inputs called \texttt{props} which can contain data that can be used in the React elements.
If a component is used in another component, we can use \texttt{props} to pass data to the child component.

> \\\\

Reviewer: ```suggestion
```

\subsection*{Props}

> When a component is instantiated it is possible to pass an arbitrary number of arguments to the component by passing key and value pairs. These are then accessed through the props object in the component. An example of this can be seen in the following listing.

Reviewer: ```suggestion
When a component is instantiated it is possible to pass an arbitrary number of arguments to the component by passing key and value pairs.
These are then accessed through the props object in the component.
An example of this can be seen in the following listing.
```

> \begin{lstlisting}[label={lst:react-props-lsting}]

Reviewer: ref til labellen herover

function Site(){
	const titleVariable = 'title';
	const descriptionVariable = 'This is a component';

	return <TextComponent title={titleVariable} description={descriptionVariable} />;
}

function TextComponent (props){
	return (
		<div>
			<h1>{props.title}</h1>
			<p>{props.description}</p>
		</div>
	);
}
\end{lstlisting}

The props of a component should be treated as read-only because any changes to the props will not be reflected in the rendering of the component.

\subsection*{State}

> There two kinds of components which are stateful and stateless components.

Reviewer: ```suggestion
There two kinds of components, which are stateful and stateless components.
```

In our application we will mostly implement stateful components as class components and stateless components as functional components. 

> Functional components can also be stateful, but the choice of using class components is mostly a question about preferences.

Reviewer: ```suggestion
Functional components can also be stateful, but the choice of using class components is mostly a question of preference.
```

We can set the initial state of a component by assigning \texttt{this.state} in the constructor of the component.
\begin{center}
    \texttt{this.state = \{textToDisplay: 'Initial text'\};}
\end{center}
The state can then be changed by reassigning to the \texttt{date} variable.
This is done by using the \texttt{this.setState()} function.
\begin{center}
    \texttt{this.setState(\{textToDisplay: 'New text'\});}
\end{center}
Parent or child components should not care if a certain component is stateful or stateless since the state is local and not accessible by other components.
A component can however pass variables from its state through \texttt{props} to a child component, which means that information about a state can only be passed in a \textit{top-down} approach \cite{ReactJS}.
Every time a state is changed with \texttt{this.setState()} the component that the state belongs to and all of its children are re-rendered.
By passing callback to the props of children we can allow  the children of a component to transfer data back to its parent and even trigger re-renders by passing callback functions that set the state of the parent component.

> If a function calls \texttt{this.setState()} it has to be bound in the constructor of the component.

Reviewer: ```suggestion
If a function calls \texttt{this.setState()}, it has to be bound in the constructor of the component.
```

> An example of how this is done can be seen in the following listing.

Reviewer: ref listing

\begin{lstlisting}
class ourComponent extends Component {
	constructor(props){
		super(props);
		this.state = {
			textToDisplay: 'initial text'
		};

		this.handleChange = this.handleChange.bind(this);
	}

	handleChange(text){
		this.setState({
			textToDisplay: text
		});
	}

	render(){
		return <ChildComponent handleChange={this.handleChange} />
	}
}
\end{lstlisting}

\subsection*{Lifecycle}
When a component is implemented as a class that extends \texttt{Component} from the react library, it gives us access to some functions that are automatically called at different points in the lifecycle of a component.

> The function that we will mostly be using is \texttt{componentDidMount}.

Reviewer: måkse tilføj en ref der lige viser funktionen og et eksempel på at vi kalder noget i den

This function is called right after the component is rendered for the first time in its lifecycle.

> This is useful for when we need to do fetch data from the API that we want to save in the state of the component.

Reviewer: ```suggestion
This is useful for when we need to fetch data from the API that we want to save in the state of the component.
```

> The reason that this is not done in the constructor of the component is because, the render function should not be waiting for asynchronous operations before it is allowed to render.

Reviewer: ```suggestion
The reason that this is not done in the constructor of the component is that the render function should not be waiting for asynchronous operations before it is allowed to render.
```

> This is because we do not know how long it will take before function is done executing or if it will even succeed.

Reviewer: ```suggestion
This is because we do not know how long it will take before the asynchronous function is done executing, or if it will even succeed.
```

> So the rendering of the component should not be waiting for this.

Reviewer: ```suggestion
```

> Instead it is better to first render the component before calling asynchronous functions that will trigger a re-rendering of the component.

Reviewer: ```suggestion
Instead of waiting, it is better to first render the component before calling asynchronous functions that will trigger a re-rendering of the component.
```
